### ISSUE TITLE
Question: Assistance with Flatpak and 5.0.0

### ISSUE THREAD

**Issue by AsavarTzeth** [wwmm/pulseeffects#894]
I recently began the work to [get 5.0.0 working with flatpak](https://github.com/flathub/com.github.wwmm.pulseeffects/tree/testing) but I am having some issues. I have already confirmed that the Arch Linux community package works on my system.

- It triggers on launch and every time I attempt to enable an application.
- No sound is processed.
- Granting all possible Flatpak permissions make no difference.

Standard output:
```
(pulseeffects:2): pulseeffects-WARNING **: 09:05:17.676: pipe_manager: Remote error on id:19

(pulseeffects:2): pulseeffects-WARNING **: 09:05:17.676: pipe_manager: Remote error message:no permission to call method 1 on 19 (requires 000000c0, have 00000140)
```

For starters I suppose I just need help interpreting this. A full debug output can be found [here](https://gist.github.com/AsavarTzeth/329845727ca74d397e1f537be004e5b8).

**Comment by wwmm** [comment]
Someone else who tried to use flatpak had problems too https://github.com/wwmm/pulseeffects/issues/397#issuecomment-757546528. I think we will need guidance from PipeWire people on that. I will open an issue there asking for help.

**Comment by wwmm** [comment]
I opened an issue there https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/667

**Comment by wwmm** [comment]
PipeWire's developer answered. It seems this will require some improvements in PipeWire. For now he suggested a workaround:

```
Meanwhile, you could add in pipewire.conf, for the access module: access.allowed=pulseeffects. I'll see if this can be made a bit better now as well.
```

Can you test if that makes things work?

**Comment by wwmm** [comment]
PipeWire master branch has already received some chagnes that will help with this situation:
https://gitlab.freedesktop.org/pipewire/pipewire/-/commit/a48a4fc1bd689a4a97d875b1d2b83adbc44e6f63
https://gitlab.freedesktop.org/pipewire/pipewire/-/commit/b8ef0fc5bbb7740ab51d7a8ce7a922760c931d66

The user will have to mnually add pulseeffects to the list of allowed applications. But it is a start.

**Comment by AsavarTzeth** [comment]
I read everything thoroughly and it was all a great read and very informative. As was your porting journey issue by the way. My compliments to the PipeWire developers for the great support.

> Can you test if that makes things work?

I added `access.allowed=pulseeffects` and success! Everything at least appears to work as usual.

> The user will have to mnually add pulseeffects to the list of allowed applications. But it is a start.

It will indeed do for now. Besides, I am going to provide this in the flathub beta branch anyhow. That means all current users already have to explicitly install it from the command line (and hopefully) inform themselves as needed.

_What I can do is add a mention of this requirement in a `README.md`, also committed on the beta branch._

**Comment by AsavarTzeth** [comment]
On one occasion I also had this spam my terminal:

    [W][000008905.960938][impl-node.c:1040 node_on_fd_events()] (pulseeffects_soe-17) client missed 1 wakeups

**Comment by wwmm** [comment]
>client missed 1 wakeups

I think that other applications also show this in some situations. This might be a PipeWire issue.

**Comment by AsavarTzeth** [comment]
I might have spoken too soon. It appeared to work once and now I cannot replicate it.

It is probably something obvious I missed. I likely just ned some rest and will try again tomorrow.

**Comment by wwmm** [comment]
> pw-cli dump still reports pipewire.access = "flatpak"

This is correct. In the issue I opened there he explained that the check and eventual upgrade to from a read-only to a write permission will be done internally.

**Comment by MrSenshi** [comment]
@AsavarTzeth Can't build, zita-convolver won't built because it's missing something.
"bin/ld: cannot find -lfftw3f
collect2: error: ld returned 1 exit status"

**Comment by AsavarTzeth** [comment]
@MrSenshi I saw your comment in my email. In case you still have that issue, have you forgotten to fetch the git submodule: `shared-modules`? Perhaps you realized that soon after and deleted the comment?

As you can [see in the manifest](https://github.com/flathub/com.github.wwmm.pulseeffects/blob/testing/com.github.wwmm.pulseeffects.yml#L179), fftw3f is included.

**Comment by MrSenshi** [comment]
@AsavarTzeth I realised my issue soon after writing the message, my mistake! Sorry for wasting your time.

**Comment by MrSenshi** [comment]
@AsavarTzeth Found a temporary workaround. Edit /etc/pipewire/media-session.d/media-session.conf to include 
```
with-flatpak = [
    with-pulseaudio
]
```
within modules, you should see similar entries for jack and alsa.

This opens it up to all flatpaks, which is not ideal, and also has the side effect of you needing to manually set the input/output of different apps that are recording/playing audio to the Pulseeffects source/sink manually using pavaucontrol. 

Tested on Alpine Linux.

**Comment by wwmm** [comment]
Wouldn't it be easier to add pulseeffects to the `access.allowed` array?

**Comment by MrSenshi** [comment]
@wwmm Couldn't get that to work.

**Comment by wwmm** [comment]
>Couldn't get that to work.

Maybe the Pipewire commits that I saw related to this were not released yet.

**Comment by AsavarTzeth** [comment]
I can now confirm that it works. But the host `pipewire` must currently be built from `master`, meaning that when pulseeffects runs in a sandbox it will depend on **pipewire >= 0.3.22**.

**Comment by AsavarTzeth** [comment]
Apparently I get these errors, on each click:

```
(pulseeffects:2): GLib-GIO-CRITICAL **: 14:34:45.142: is_key: assertion 'key[length - 1] != '/'' failed
(pulseeffects:2): GLib-GIO-CRITICAL **: 14:34:45.142: g_settings_backend_changed: assertion 'is_key (key)' failed
```

It seems like there is an issue with the dconf keyfile backend.

**Comment by wwmm** [comment]
>Not sure, why it didn't work with a regular reset in PulseEffects.

Strange indeed. Based on the `is_key` error the way we are handling this is not playing nice with flatpak. What we do is caling the reset method on an empty string https://github.com/wwmm/pulseeffects/blob/d18451b36c0b7a8c6298f97cb1b3ddb0f0ee5059/src/general_settings_ui.cpp#L206. This forces a recursive reset that clears all the keys. It works fine outside of Flatpak.

**Comment by AsavarTzeth** [comment]
Could you try with `GSETTINGS_BACKEND=keyfile`? I do believe that should work outside of flatpaks too.

I don't remember this being an issue back when I switched the back-end. Although neither can I say for sure that it's a recent development.

A good explanation of the backend migration:
- https://blogs.gnome.org/mclasen/2019/07/12/settings-in-a-sandbox-world/

**Comment by wwmm** [comment]
>Could you try with GSETTINGS_BACKEND=keyfile?

Now I see the same crash.

**Comment by wwmm** [comment]
I took a look at a few glib backend files https://github.com/GNOME/glib/tree/master/gio and different backend have different reset approach. 

I am not sure if we can fix this on PulseEffects without adding lots of code. Having to reset each key individually is really annoying...

**Comment by AsavarTzeth** [comment]
Wow, I had no idea. It really is unfortunate if indeed 5.0.0 happens to require a reset (if that wasn't exclusive to me). Since 5.x will live in the beta branch for quite some time, at least there's no rush to figure it out.

**Comment by AsavarTzeth** [comment]
Here's a thought. Assuming we have a reliable and not hardcoded way of finding the key file, could we simply delete it (and then offer to restart the app)? You could detect the current gsettings backend and only do this when a key file is used.

It does seem a bit like a hack though... Just trying to think outside the box.

**Comment by wwmm** [comment]
>if indeed the update to 5.0.0 happens to require a reset 

This is not clear yet. I still do not understand why a reset was necessary. There weren't many changes related to the database keys. Some keys were removed and I think only one was added. Nothing huge...

>Assuming we have a reliable and not hardcoded way of finding the key file, could we simply delete it (and then offer to restart the app)?

As far as I remember it is not so simple. For performance reasons gsettings handles everything in a binary file that its server handles. Even if editing the database file directly were a viable solution gsettings server has to be notified about it for applications to see the change. I think we would be fighting against its server. The real solution here would be an improved reset function in the keyfile backend. 

>Another non-proper idea is something like gsettings reset-recursively com.github.wwmm.pulseeffects

The call to `settings->reset("")` does the recursive reset when the default backend is used.But for some reason `gsettings reset-recursively com.github.wwmm.pulseeffects` does not reset all the keys even when the default backend is used. So it won't solve the problem.
